### PR TITLE
tlt-3051: isites_migration - migrations now use S3

### PR DESCRIPTION
**Pull request by elliottyates**
`isites_migration` no longer accesses fsdocs via file system; uses S3 instead, and can be run entirely from EC2 virtualenv. After conferring with Devops, it no longer uses an AWS profile loaded on the machine, but rather credentials stored in the secure.py. We could change this later to use a profile that was updated at deployment time, or the IAM role associated with the machine.

To deploy this seamlessly, needs https://github.com/Harvard-University-iCommons/tlt-ops/pull/244.

# Testing:
This will not work as expected in our dev, qa, and stage environments as the fsdocs we have in S3 are prod filenodes, so the keyword mappings with dev, qa, and stage won't be correct. I tested this locally by changing the `get_previous_isites()` call in ` isites_migration.views.index` to simply return the list I wanted, e.g. `[{'term': u'test', 'calendar_year': 2017, 'keyword': u'k102653', 'title': u'Test'}]`.

After discussing with PO and QA, we decided that we won't do any extra work to make this QA'able for the moment. Right now you _could_ QA this by installing on our beta or test instance and changing the DB temporarily to prod, and _only_ testing this app. Otherwise we'd want to push this to prod and test it there. Some other possibilities were explored:
- Moving QA fsdocs data into S3; and,
- changing the particular calls we need for the file system data to `.using()` the prod db (and finding 'magic courses' where the QA keyword matched a useful set of files to test against in a prod site with the same keyword), but this would expose prod credentials in our non-prod S3 settings buckets.

Note that unit tests are passing but selenium tests have been failing for some time (see [TLT-3062](https://jira.huit.harvard.edu/browse/TLT-3062)).

## secure.py requirements for testing locally:

The following will be needed in secure.py to test this locally (use AWS credentials that can read/write from the specified buckets):
```
    'isites_migration_aws_access_key_id': '',
    'isites_migration_aws_secret_access_key': '',
    'aws_export_bucket_isites_files': 'isites-exports-non-prod',
    'aws_source_bucket_isites_files': 'uw-isites-fsdocs-archive-dev',
```

### CODE REVIEW

File: canvas_manage_course/settings/isites_export_tool.py [hunk 1]
```diff
@@ -1,15 +1,24 @@
 from .aws import *
 
 INSTALLED_APPS = (
+    'async',
```
Comment (elliottyates) [inline canvas_manage_course/settings/isites_export_tool.py]: As discussed with @sapnamysore offline, the old version of this stuff, in canvas_course_admin_tools, didn't make have lti_school_permissions, manage_people, or manage_sections. Without them in INSTALLED_APPS, Django fails to process canvas_manage_course/urls.py when a command is run under these settings. (Not sure why `async` was also causing a failure here, possibly because of a move from Django 1.8 to 1.9?) If there is a more elegant way to avoid these apps that aren't germane to a command, would be good to know.

File: canvas_manage_course/settings/isites_migration.py [hunk 2]
```diff
@@ -1,17 +1,20 @@
...
+    'manage_people',
+    'manage_sections',
 )
 
 # Settings specifically for the isite_migration LTI app
 AWS_EXPORT_BUCKET_ISITES_FILES = 'isites-slide-data'
 AWS_EXPORT_BUCKET_SLIDE_TOOL = 'isites-slide-data'
 AWS_EXPORT_DOWNLOAD_TIMEOUT_SECONDS = 60
-AWS_PROFILE = 'isites_migration'
+AWS_SOURCE_BUCKET_ISITES_FILES = SECURE_SETTINGS.get('aws_source_bucket_isites_files', 'uw-isites-fsdocs-archive')
```
Comment (elliottyates) [inline canvas_manage_course/settings/isites_migration.py]: This default bucket, `uw-isites-fsdocs-archive`, doesn't actually exist right now. @cmurtaugh, do we expect a 'prod' version of this bucket to be called this? Should we default to the `-dev` version? Or should we just ensure that the right secure variables get set to point to `-dev` until such time we rename the bucket?

**Comment by elliottyates** [comment]
Another thing to keep in mind, @cmurtaugh will we need to refresh fsdocs in s3 from time to time?

File: canvas_manage_course/settings/isites_export_tool.py (hunk 1 above)
Comment (sapnamysore) [inline canvas_manage_course/settings/isites_export_tool.py]: Thanks for documenting why we needed it

File: isites_migration/management/commands/export_slide_tool.py [hunk 3]
```diff
@@ -67,29 +81,55 @@ def handle(self, *args, **options):
...
                         logger.error("Failed to find storage node for file node %d", file_node.file_node_id)
                         continue
 
-                    file_name = os.path.basename(file_node.physical_location)
-                    s3_key = "%s/%d/%s" % (keyword, topic.topic_id, file_name)
+                    # Determine if we're dealing with an fs-docs storage node or an fs-cow
+                    if 'fs-docs' in storage_node_loc:
+                        # For pre-COW files, we need to construct the physical file location
+                        # ourselves
+                        physical_file_loc = os.path.join(
```
Comment (sapnamysore) [inline isites_migration/management/commands/export_slide_tool.py]: Thanks for clarifying that this logic came from the predecessor of this repo (canavas_course_admin_tool)

File: isites_migration/management/commands/export_slide_tool.py [hunk 4]
```diff
@@ -67,29 +81,55 @@ def handle(self, *args, **options):
...
+                    ).encode('utf8')
+
+                    source_object_key = source_file[source_file.find('/fsdocs') + 1:]
+
+                    copy_source = {
+                        'Bucket': s3_source_bucket,
+                        'Key': source_object_key
+                    }
+
+                    # storage_node_loc + file_node.physical_location,
```
Comment (sapnamysore) [inline isites_migration/management/commands/export_slide_tool.py]: meant as a comment?

**Comment by sapnamysore** [comment]
👍

File: isites_migration/management/commands/export_slide_tool.py (hunk 4 above)
Comment (elliottyates) [inline isites_migration/management/commands/export_slide_tool.py]: cruft! Will remove it. :)

File: isites_migration/views.py [hunk 5]
```diff
@@ -6,7 +6,6 @@
 
 from lti_school_permissions.decorators import lti_permission_required
 from async.models import Process
```
Comment (elliottyates) [inline isites_migration/views.py]: Sure -- was just trying to minimize the refactoring, but happy to reorganize here..

File: canvas_manage_course/settings/isites_migration.py [hunk 6]
```diff
@@ -1,17 +1,32 @@
...
+"""
 from .aws import *
 
 INSTALLED_APPS = (
-    'icommons_common',
     'async',
     'django_rq',
+    'icommons_common',
     'isites_migration',
+    'lti_school_permissions',
```
Comment (bermudezjd) [inline canvas_manage_course/settings/isites_migration.py]: Don't mean to dredge up an already approved PR, but I was looking through the code to see if there were any new/modified secure settings files and I have a question about these installed apps.  When using this settings file, why do we now need `manage_people`, `manage_sections`, and `lti_school_permissions`?  Is this related to unit tests?

Comment (elliottyates) [inline canvas_manage_course/settings/isites_migration.py]: IIRC suddenly you couldn't run the commands with these settings anymore using manage.py. I think it was something to do with parsing URL includes (the app name was not in the installed apps and it failed?). I can't remember the details, but happy to troubleshoot if needed.

Comment (bermudezjd) [inline canvas_manage_course/settings/isites_migration.py]: Hmmm, I don't see any updates to url configurations with this PR.  You're sure it was failing without these apps?

Comment (elliottyates) [inline canvas_manage_course/settings/isites_migration.py]: I'm sure it was, but I don't know why. This fixed it, and I didn't investigate further due to the priority of these features. I would be interested to know why, but didn't follow up, sorry. It would take me awhile to switch back over to the tlt-vagrant VM setup and re-test this, but we could pull these apps out and run the management commands with these settings to see if there's another way to get around the issue, or determine exactly what the issue is.

Comment (elliottyates) [inline canvas_manage_course/settings/isites_migration.py]: Looks like this is part of moving up to Django 1.9, which was part of the most recent release. https://github.com/Harvard-University-iCommons/canvas_manage_course/commit/12dab87d3cd1f9bc5acb6d5c47c2322dc69ab7ad

The issue appears to be that the Django startup checks now fail if models are referenced in apps that aren't part of `INSTALLED_APPS` and don't have `app_label`s in their model `Meta:` data.
https://docs.djangoproject.com/en/1.10/ref/models/options/#app-label

I verified that adding `app_label`s fixes this issue and will submit PRs to clean this up.

Here's an example stack trace of the issue once you move to Django 1.9,  don't have these in `INSTALLED_APPS`, and they don't have `app_label`s.

```
Traceback (most recent call last):
File "/home/vagrant/tlt/canvas_manage_course/manage.py", line 13, in <module>
execute_from_command_line(sys.argv)
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/django/core/management/init.py", line 353, in execute_from_command_line
utility.execute()
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/django/core/management/init.py", line 345, in execute
self.fetch_command(subcommand).run_from_argv(self.argv)
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/django/core/management/base.py", line 348, in run_from_argv
self.execute(*args, **cmd_options)
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/django/core/management/base.py", line 398, in execute
self.check()
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/django/core/management/base.py", line 426, in check
include_deployment_checks=include_deployment_checks,
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/django/core/checks/registry.py", line 75, in run_checks
new_errors = check(app_configs=app_configs)
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/django/core/checks/urls.py", line 13, in check_url_config
return check_resolver(resolver)
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/django/core/checks/urls.py", line 23, in check_resolver
for pattern in resolver.url_patterns: File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/django/utils/functional.py", line 33, in get
res = instance.dict[self.name] = self.func(instance)
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/django/core/urlresolvers.py", line 417, in url_patterns
patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/django/utils/functional.py", line 33, in get
res = instance.dict[self.name] = self.func(instance)
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/django/core/urlresolvers.py", line 410, in urlconf_module
return import_module(self.urlconf_name)
File "/usr/lib/python2.7/importlib/init.py", line 37, in import_module
import(name)
File "/home/vagrant/tlt/canvas_manage_course/canvas_manage_course/urls.py", line 3, in <module>
from canvas_manage_course import views
File "/home/vagrant/tlt/canvas_manage_course/canvas_manage_course/views.py", line 15, in <module>
from lti_school_permissions.decorators import (
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/lti_school_permissions/decorators.py", line 8, in <module>
from lti_school_permissions.verification import is_allowed
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/lti_school_permissions/verification.py", line 12, in <module>
from lti_school_permissions.models import SchoolPermission
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/lti_school_permissions/models.py", line 7, in <module>
class SchoolPermission(models.Model):
File "/home/vagrant/.virtualenvs/canvas_manage_course/local/lib/python2.7/site-packages/django/db/models/base.py", line 102, in new
"INSTALLED_APPS." % (module, name)
RuntimeError: Model class lti_school_permissions.models.SchoolPermission doesn't declare an explicit app_label and isn't in an application in INSTALLED_APPS.
```
